cli/vscodeipc/vscodeipc.go

```diff
@@ -0,0 +1,298 @@
+		regionParts := strings.SplitN(rawRegion, "-", 2)
+		regionID, err := strconv.Atoi(regionParts[0])
+		if err != nil {
+			continue
+		}
+		region, found := derpMap.Regions[regionID]
+		if !found {
+			// It's possible that a workspace agent is using an old DERPMap
+			// and reports regions that do not exist. If that's the case,
+			// report the region as unknown!
+			region = &tailcfg.DERPRegion{
+				RegionID:   regionID,
+				RegionName: fmt.Sprintf("Unnamed %d", regionID),
+			}
+		}
+		// Convert the microseconds to milliseconds.
+		derpLatency[region.RegionName] = latency * 1000
+	}
+
+	totalRx := uint64(0)
+	totalTx := uint64(0)
+	for _, stat := range api.agentConn.ExtractTrafficStats() {
+		totalRx += stat.RxBytes
+		totalTx += stat.TxBytes
+	}
+	// Tracking the time since last request is required because
+	// ExtractTrafficStats() resets its counters after each call.
+	dur := time.Since(api.lastNetwork)
+	uploadSecs := float64(totalTx) / dur.Seconds()
+	downloadSecs := float64(totalRx) / dur.Seconds()
+
+	api.lastNetwork = time.Now()
+
+	httpapi.Write(r.Context(), w, http.StatusOK, NetworkResponse{
+		P2P:              p2p,
+		Latency:          float64(latency.Microseconds()) / 1000,
+		PreferredDERP:    derpMap.Regions[node.PreferredDERP].RegionName,
+		DERPLatency:      derpLatency,
+		UploadBytesSec:   int64(uploadSecs),
+		DownloadBytesSec: int64(downloadSecs),
+	})
+}
+
+type ExecuteRequest struct {
+	Command string `json:"command"`
+}
+
+type ExecuteResponse struct {
+	Data     string `json:"data"`
+	ExitCode *int   `json:"exit_code"`
+}
+
+// execute runs the command provided, streams the output back, and returns the exit code.
+func (api *api) execute(w http.ResponseWriter, r *http.Request) {
+	var req ExecuteRequest
+	if !httpapi.Read(r.Context(), w, r, &req) {
+		return
+	}
+	api.sshClientOnce.Do(func() {
+		// The SSH client is lazily created because it's not needed for
+		// all requests. It's only needed for the execute endpoint.
+		api.sshClient, api.sshClientErr = api.agentConn.SSHClient(context.Background())
+	})
+	if api.sshClientErr != nil {
+		httpapi.Write(r.Context(), w, http.StatusInternalServerError, codersdk.Response{
+			Message: "Failed to create SSH client.",
+			Detail:  api.sshClientErr.Error(),
+		})
+		return
+	}
+	session, err := api.sshClient.NewSession()
+	if err != nil {
+		httpapi.Write(r.Context(), w, http.StatusInternalServerError, codersdk.Response{
+			Message: "Failed to create SSH session.",
+			Detail:  err.Error(),
+		})
+		return
+	}
+	defer session.Close()
+	f, ok := w.(http.Flusher)
+	if !ok {
+		httpapi.Write(r.Context(), w, http.StatusInternalServerError, codersdk.Response{
+			Message: fmt.Sprintf("http.ResponseWriter is not http.Flusher: %T", w),
+		})
+		return
+	}
+
+	execWriter := &execWriter{w, f}
+	session.Stdout = execWriter
+	session.Stderr = execWriter
+	session.Stdin = strings.NewReader(req.Command + "\n")
+	err = session.Start("sh")
```

Comment:
Do we want to use `sh` or user shell here? Windows compat?

Reply:
Fixed!